File: app/src/main/java/com/voipgrid/vialer/call/CallActionButton.java
package com.voipgrid.vialer.call;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.core.content.res.ResourcesCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import android.util.AttributeSet;

import com.voipgrid.vialer.R;

public class CallActionButton extends AppCompatImageView {

    public static final float ENABLED_ALPHA = 1.0f, DISABLED_ALPHA = 0.5f;

    public CallActionButton(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void enable() {
        this.setActivated(true);
        this.setAlpha(ENABLED_ALPHA);
    }

    public void disable() {
        this.setActivated(false);
        this.setAlpha(DISABLED_ALPHA);
    }

    public void enable(boolean enable) {
        if (enable) {
            this.enable();
        } else {
            this.disable();
        }
    }

    public void activate() {
        this.setColorFilter(ResourcesCompat.getColor(getResources(), R.color.color_primary, null));
        Drawable newDrawableSpeaker = this.getBackground();
        DrawableCompat.setTint(newDrawableSpeaker, Color.WHITE);
        this.setBackground(newDrawableSpeaker);
    }

    public void deactivate() {
        this.clearColorFilter();
        Drawable newDrawableSpeaker = this.getBackground();
        DrawableCompat.setTintList(newDrawableSpeaker, null);
        this.setBackground(newDrawableSpeaker);
    }

    public void activate(boolean activate) {
        if (activate) {
            this.activate();
        } else {
            this.deactivate();
        }
    }
}

File: app/src/main/java/com/voipgrid/vialer/call/CallActionButton.kt
package com.voipgrid.vialer.call

import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.util.Log
import android.widget.LinearLayout
import androidx.core.graphics.drawable.DrawableCompat
import com.voipgrid.vialer.R
import kotlinx.android.synthetic.main.call_button.view.*

open class CallActionButton(context: Context, attrs: AttributeSet?) : LinearLayout(context, attrs) {

    private val drawable: Drawable?
    private val originalDrawable: Drawable
    private val originalText: String

    init {
        inflate(context, R.layout.call_button, this)
        val attributes = context.obtainStyledAttributes(attrs, R.styleable.CallActionButton)
        label.text = attributes.getString(R.styleable.CallActionButton_label)
        image.setImageDrawable(attributes.getDrawable(R.styleable.CallActionButton_image))
        drawable = attributes.getDrawable(R.styleable.CallActionButton_image)
        attributes.recycle()
        originalDrawable = image.drawable
        originalText = label.text.toString()
    }

    fun enable() {
        this.isActivated = true
        this.alpha = ENABLED_ALPHA
    }

    fun disable() {
        this.isActivated = false
        this.alpha = DISABLED_ALPHA
    }

    fun activate() {
        image.setColorFilter(Color.WHITE)
        image.background.setColorFilter(resources.getColor(R.color.color_primary), PorterDuff.Mode.SRC_ATOP)
        label.setTextColor(resources.getColor(R.color.color_primary))
    }

    fun deactivate() {
        image.clearColorFilter()
        image.background.clearColorFilter()
        label.setTextColor(resources.getColor(R.color.call_color))
    }

    fun swapIconAndText(replacementImage: Int, replacementText: Int) {
        image.setImageResource(replacementImage)
        label.text = context.getString(replacementText)
    }

    fun resetIconAndText() {
        if (image.drawable != originalDrawable) {
            image.setImageDrawable(originalDrawable)
        }

        label.text = originalText
    }

    companion object {

        const val ENABLED_ALPHA = 1.0f
        const val DISABLED_ALPHA = 0.5f
    }

    fun enable(enable: Boolean): Unit = when (enable) {
        true -> this.enable()
        false -> this.disable()
    }


    fun activate(activate: Boolean): Unit = when (activate) {
        true -> this.activate()
        false -> this.deactivate()
    }
}
